README.md

<h1 align="center">
  <br>
  <a href="https://korolvs.github.com/resting_pug"><img src="https://korolvs.github.com/resting_pug/logo.png" alt="Logo" width="80%"></a>
  <br>
</h1>

[![Gem Version](https://badge.fury.io/rb/resting_pug.svg)](https://badge.fury.io/rb/resting_pug)
[![Maintainability](https://api.codeclimate.com/v1/badges/e53c62ffa3ee4cba3895/maintainability)](https://codeclimate.com/github/korolvs/resting_pug/maintainability)
[![Build Status](https://travis-ci.org/korolvs/resting_pug.svg?branch=master)](https://travis-ci.org/korolvs/resting_pug)
[![codecov](https://codecov.io/gh/korolvs/resting_pug/branch/master/graph/badge.svg)](https://codecov.io/gh/korolvs/resting_pug)

## Table of Contents

- [What is Resting Pug?](#what-is-resting-pug)
- [Why should I use it?](#why-should-i-use-it)
- [Installation](#installation)
- [Usage](#usage)
- [Created actions and thing that you can customize](#created-actions-and-things-you-can-customize)
  - [Create](#create)
  - [Update](#update)
  - [Destroy](#destroy)
  - [Show](#show)
  - [Index](#index)
- [Contributing](#contributing)
- [Changelog](#changelog)
- [License](#license)

## What is Resting Pug?
Resting Pug allows you to create a JSON API with just adding one line to your controller and fully customize it with overriding small and simple helper methods.

## Why should I use it?
Often when you want to try a new idea or create a simple app you want to implement basic features very fast.
Resting Pug allows you to do it with just couple lines of code.
But even though it is small and simple(and because it is small and simple) you can vastly customize created API.
Attributes user can see, attributes user can edit, number of items on page, how to render items and errors - thats just a small part of things you can change overriding basic Resting Pug methods.

## Installation
Add this line to your application's Gemfile:

```ruby
gem 'resting_pug'
```

And then execute:
```bash
$ bundle
```

## Usage

**1. Create a model that will be accesible through API:**

```ruby
class Book < ApplicationRecord
  validates :title, presence: true, length: { minimum: 3 }
  validates :author, presence: true, length: { minimum: 3 }
end
```

```ruby
class CreateBooks < ActiveRecord::Migration
  def change
    create_table :books do |t|
      t.string :title
      t.string :author
      t.integer :year
      t.integer :rating

      t.timestamps
    end
  end
end
```

**2. Create a new controller and include ```RestingPug::Base``` into it to add CRUD actions:**

```ruby
class BooksController < ApplicationController
  include RestingPug::Base
end
```

You can add CRUD actions to all controllers by including this module into ```ApplicationController```:

```ruby
class ApplicationController < ActionController::Base
  include RestingPug::Base
end
```

**3. Don't forget to add a route to ```config/routes.rb```**

```ruby
Rails.application.routes.draw do
  resources :books
end
```

**4. Enjoy your new API and don't hesitate to change it whatever you want!**

## List of actions and things that you can customize
Actions created after including ```RestingPug::Base```:

### Create

```
Request:
POST http://your.awesome/api/books
{
  "name": "11/22/63",
  "author": "Stephen King"
}

Response:
200 OK
{
  "book": {
    "id": 1,
    "name": "11/22/63",
    "author": "Stephen King"
  }
}
```

Things you can customize:

- override [permitted_fields_for_create](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_create) to set which params can be set in a new subject
- override [permitted_fields_for_show](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_show) to set which params will be shown in response
- override [render_subject](https://korolvs.github.com/resting_pug/0.1.1/render/render_subject) to set how subject will be rendered
- override [render_errors](https://korolvs.github.com/resting_pug/0.1.1/render/render_errors) to set how errors will be rendered
- override [subject_model](https://korolvs.github.com/resting_pug/0.1.1/subject/subject_model) to set what model will be created
- override [create_chain](https://korolvs.github.com/resting_pug/0.1.1/chains/create_chain) to add or remove methods which will be called while creating a subject
- override [decide_what_to_render](https://korolvs.github.com/resting_pug/0.1.1/render/decide_what_to_render) to set how it will be decided what to render in response

### Update

```
Request:
PATCH http://your.awesome/api/books/1
{
  "name": "The Green Mile",
}

Response:
200 OK
{
  "book": {
    "id": 1,
    "name": "The Green Mile",
    "author": "Stephen King"
  }
}
```

Things you can customize:

- override [permitted_fields_for_update](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_update) to set which params can be updated
- override [permitted_fields_for_show](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_show) to set which params will be shown in response
- override [render_subject](https://korolvs.github.com/resting_pug/0.1.1/render/render_subject) to set how subject will be rendered
- override [render_errors](https://korolvs.github.com/resting_pug/0.1.1/render/render_errors) to set how errors will be rendered
- override [render_not_found](https://korolvs.github.com/resting_pug/0.1.1/render/render_not_found) to set what to render when subject with ID from params is not found
- override [subject_model](https://korolvs.github.com/resting_pug/0.1.1/subject/subject_model) to set what model will be updated
- override [update_chain](https://korolvs.github.com/resting_pug/0.1.1/chains/update_chain) to add or remove methods which will be called while updating a subject
- override [decide_what_to_render](https://korolvs.github.com/resting_pug/0.1.1/render/decide_what_to_render) to set how it will be decided what to render in response

### Destroy

```
Request:
DELETE http://your.awesome/api/books/1

Response:
204 No Content
```

Things you can customize:

- override [destroy_subject](https://korolvs.github.com/resting_pug/0.1.1/subject/destroy_subject) to set how it will be destroyed
- override [render_nothing](https://korolvs.github.com/resting_pug/0.1.1/render/render_nothing) to set what to render when subject is destroyed
- override [render_errors](https://korolvs.github.com/resting_pug/0.1.1/render/render_errors) to set how errors will be rendered
- override [render_not_found](https://korolvs.github.com/resting_pug/0.1.1/render/render_not_found) to set what to render when subject with ID from params is not found
- override [subject_model](https://korolvs.github.com/resting_pug/0.1.1/subject/subject_model) to set what model will be deleted
- override [destroy_chain](https://korolvs.github.com/resting_pug/0.1.1/chains/destroy_chain) to add or remove methods which will be called while deleting a subject

### Show

```
Request:
GET http://your.awesome/api/books/1

Response:
200 OK
{
  "book": {
    "id": 1,
    "name": "The Green Mile",
    "author": "Stephen King"
  }
}
```

Things you can customize:

- override [permitted_fields_for_show](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_show) to set which params will be shown in response
- override [render_subject](https://korolvs.github.com/resting_pug/0.1.1/render/render_subject) to set how subject will be rendered
- override [render_not_found](https://korolvs.github.com/resting_pug/0.1.1/render/render_not_found) to set what to render when subject with ID from params is not found
- override [subject_model](https://korolvs.github.com/resting_pug/0.1.1/subject/subject_model) to set what model will be shown
- override [show_chain](https://korolvs.github.com/resting_pug/0.1.1/chains/show_chain) to add or remove methods which will be called while updating a subject

### Index

```
Request:
GET http://your.awesome/api/books
GET http://your.awesome/api/books?filter[author]="King"
GET http://your.awesome/api/books?filter[author][]="King"&filter[author][]="Kesey"
GET http://your.awesome/api/books?sort=-id,name,-author
GET http://your.awesome/api/books?page=3&per_page=10
GET http://your.awesome/api/books?filter[author]="King"&sort=-id?page=3&per_page=10

Response:
200 OK
{
  "books": [{
    "id": 2,
    "name": "The Green Mile",
    "author": "Stephen King"
  }, {
    "id": 1,
    "name": "11/22/63",
    "author": "Stephen King"
  }]
}
```

Things you can customize:

- override [permitted_fields_for_show](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_show) to set which params will be shown in response
- override [permitted_fields_for_filter](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_filter) to set which params can be used for filtering
- override [permitted_fields_for_sort](https://korolvs.github.com/resting_pug/0.1.1/params/permitted_fields_for_sort) to set which params can be used for sorting
- override [default_sort_params](https://korolvs.github.com/resting_pug/0.1.1/params/default_sort_params) to set default sort params
- override [per_page_default](https://korolvs.github.com/resting_pug/0.1.1/params/per_page_default) to set default per_page param
- override [per_page_range](https://korolvs.github.com/resting_pug/0.1.1/params/per_page_range) to set minimum and maximum possible per_page value
- override [render_subjects](https://korolvs.github.com/resting_pug/0.1.1/render/render_subjects) to set how subjects will be rendered
- override [subject_model](https://korolvs.github.com/resting_pug/0.1.1/subject/subject_model) to set what model will be shown
- override [index_chain](https://korolvs.github.com/resting_pug/0.1.1/chains/index_chain) to add or remove methods which will be called while updating a subject

## Contributing
You're encouraged to submit [pull requests](https://github.com/korolvs/resting_pug/pulls), [propose features and discuss issues](https://github.com/korolvs/resting_pug/issues).

See [CONTRIBUTING](CONTRIBUTING.md)

## Changelog
To see what has changed in recent versions of Resting Pug, see the [CHANGELOG](CHANGELOG.md).

## License
The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

Logo is [Designed by Freepik](https://www.freepik.com/free-vector/happy-pug-resting-on-the-sofa_1357299.htm)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Next Release
### Changed
  - Logo format to PNG

### Fixed
  - Link to Actions on Welcome page in docs

---

## [0.1.1] - 2017-11-12
### Added
  - Changelog
  - Jekyll Documentation website
  - Contibuting page
  - Bages
  - Better README text
  - Examples to documentation
  - Travis support
  - Codeclimate support
  - Codecov support

### Changed
  - Logo format to SVG

---

## [0.1.0] - 2017-11-02
Initial public release



lib/resting_pug/actions.rb
module RestingPug
  # Describes basic CRUD actions.
  # Each action is represented by a chain of methods definied in {Chains} module.
  # @see https://en.wikipedia.org/wiki/Create,_read,_update_and_delete
  # @used_in {Base.included}
  module Actions
    # Creates a subject with attributes from params and returns it.
    # @example
    #   Request:
    #   POST http://your.awesome/api/books
    #   {
    #     "name": "11/22/63",
    #     "author": "Stephen King"
    #   }
    #
    #   Response:
    #   200 OK
    #   {
    #     "book": {
    #       "id": 1,
    #       "name": "11/22/63",
    #       "author": "Stephen King"
    #     }
    #   }
    # @note You can sustomize creating action:
    #   - override {Params#permitted_fields_for_create permitted_fields_for_create} to set which params can be set in a new subject
    #   - override {Params#permitted_fields_for_show permitted_fields_for_show} to set which params will be shown in response
    #   - override {Render#render_subject render_subject} to set how subject will be rendered
    #   - override {Render#render_errors render_errors} to set how errors will be rendered
    #   - override {Subject#subject_model subject_model} to set what model will be created
    #   - override {Chains#create_chain create_chain} to add or remove methods which will be called while creating a subject
    #   - override {Render#decide_what_to_render decide_what_to_render} to set how it will be decided what to render in response
    # @use {#run_chain}
    # @use {Chains#create_chain}
    def create
      run_chain(create_chain)
    end

    # Updates a subject by id with attributes from params and returns it.
    # @example
    #   Request:
    #   PATCH http://your.awesome/api/books/1
    #   {
    #     "name": "The Green Mile",
    #   }
    #
    #   Response:
    #   200 OK
    #   {
    #     "book": {
    #       "id": 1,
    #       "name": "The Green Mile",
    #       "author": "Stephen King"
    #     }
    #   }
    # @note You can sustomize creating action:
    #   - override {Params#permitted_fields_for_update permitted_fields_for_update} to set which params can be updated
    #   - override {Params#permitted_fields_for_show permitted_fields_for_show} to set which params will be shown in response
    #   - override {Render#render_subject render_subject} to set how subject will be rendered
    #   - override {Render#render_errors render_errors} to set how errors will be rendered
    #   - override {Render#render_not_found render_not_found} to set what to render when subject with ID from params is not found
    #   - override {Subject#subject_model subject_model} to set what model will be updated
    #   - override {Chains#update_chain update_chain} to add or remove methods which will be called while updating a subject
    #   - override {Render#decide_what_to_render decide_what_to_render} to set how it will be decided what to render in response
    # @use {#run_chain}
    # @use {Chains#update_chain}
    def update
      run_chain(update_chain)
    end

    # Destroys a subject by id
    # @example
    #   Request:
    #   DELETE http://your.awesome/api/books/1
    #
    #   Response:
    #   204 No Content
    # @note You can sustomize creating action:
    #   - override {Subject#destroy_subject destroy_subject} to set how it will be destroyed
    #   - override {Render#render_nothing render_nothing} to set what to render when subject is destroyed
    #   - override {Render#render_errors render_errors} to set how errors will be rendered
    #   - override {Render#render_not_found render_not_found} to set what to render when subject with ID from params is not found
    #   - override {Subject#subject_model subject_model} to set what model will be deleted
    #   - override {Chains#destroy_chain destroy_chain} to add or remove methods which will be called while deleting a subject
    # @use {#run_chain}
    # @use {Chains#destroy_chain}
    def destroy
      run_chain(destroy_chain)
    end

    # Shows a subject by id
    # @example
    #   Request:
    #   GET http://your.awesome/api/books/1
    #
    #   Response:
    #   200 OK
    #   {
    #     "book": {
    #       "id": 1,
    #       "name": "The Green Mile",
    #       "author": "Stephen King"
    #     }
    #   }
    # @note You can sustomize creating action:
    #   - override {Params#permitted_fields_for_show permitted_fields_for_show} to set which params will be shown in response
    #   - override {Render#render_subject render_subject} to set how subject will be rendered
    #   - override {Render#render_not_found render_not_found} to set what to render when subject with ID from params is not found
    #   - override {Subject#subject_model subject_model} to set what model will be shown
    #   - override {Chains#show_chain show_chain} to add or remove methods which will be called while updating a subject
    # @use {#run_chain}
    # @use {Chains#show_chain}
    def show
      run_chain(show_chain)
    end

    # Shows a list of subjects.
    # You can filter subjects by attributes using *filter* param.
    # You can sort subjects using *sort* param. It is a string of comma-separated attributes with minus before attributes you wants to sort descending.
    # You can paginate subjcts with *page* and *per_page* params.
    # @example
    #   Request:
    #   GET http://your.awesome/api/books
    #   GET http://your.awesome/api/books?filter[author]="King"
    #   GET http://your.awesome/api/books?filter[author][]="King"&filter[author][]="Kesey"
    #   GET http://your.awesome/api/books?sort=-id,name,-author
    #   GET http://your.awesome/api/books?page=3&per_page=10
    #   GET http://your.awesome/api/books?filter[author]="King"&sort=-id?page=3&per_page=10
    #
    #   Response:
    #   200 OK
    #   {
    #     "books": [{
    #       "id": 2,
    #       "name": "The Green Mile",
    #       "author": "Stephen King"
    #     }, {
    #       "id": 1,
    #       "name": "11/22/63",
    #       "author": "Stephen King"
    #     }]
    #   }
    # @note You can sustomize creating action:
    #   - override {Params#permitted_fields_for_show permitted_fields_for_show} to set which params will be shown in response
    #   - override {Params#permitted_fields_for_filter permitted_fields_for_filter} to set which params can be used for filtering
    #   - override {Params#permitted_fields_for_sort permitted_fields_for_sort} to set which params can be used for sorting
    #   - override {Params#default_sort_params default_sort_params} to set default sort params
    #   - override {Params#per_page_default per_page_default} to set default per_page param
    #   - override {Params#per_page_range per_page_range} to set minimum and maximum possible per_page value
    #   - override {Render#render_subjects render_subjects} to set how subjects will be rendered
    #   - override {Subject#subject_model subject_model} to set what model will be shown
    #   - override {Chains#index_chain index_chain} to add or remove methods which will be called while updating a subject
    # @use {#run_chain}
    # @use {Chains#index_chain}
    def index
      run_chain(index_chain)
    end

    protected

    # Executes each method in a chain one by one
    # @param chain [Array] array of methods to execute
    # @used_in {#create}
    # @used_in {#update}
    # @used_in {#destroy}
    # @used_in {#show}
    # @used_in {#index}
    def run_chain(chain)
      chain.each do |action|
        self.send(action)
      end
    end
  end
end